File: Analysis/PersonalRecordAnalyzer.swift
import Foundation

/// Tracks rolling personal records, milestones, and activity streaks
struct PersonalRecordAnalyzer {

    private struct TrackedMetric {
        let metric: HealthMetric
        let lowerIsBetter: Bool
        let milestoneThreshold: Double?
        let milestoneLabel: String?
        let streakThreshold: Double?
        let streakLabel: String?
    }

    private static let trackedMetrics: [TrackedMetric] = [
        TrackedMetric(metric: .steps, lowerIsBetter: false,
                      milestoneThreshold: 10000, milestoneLabel: Copy.Analysis.PersonalRecord.tenKStepDay,
                      streakThreshold: 8000, streakLabel: Copy.Analysis.PersonalRecord.eightKPlusSteps),
        TrackedMetric(metric: .activeCalories, lowerIsBetter: false,
                      milestoneThreshold: nil, milestoneLabel: nil,
                      streakThreshold: nil, streakLabel: nil),
        TrackedMetric(metric: .exerciseMinutes, lowerIsBetter: false,
                      milestoneThreshold: nil, milestoneLabel: nil,
                      streakThreshold: 30, streakLabel: Copy.Analysis.PersonalRecord.thirtyPlusMinExercise),
        TrackedMetric(metric: .vo2Max, lowerIsBetter: false,
                      milestoneThreshold: nil, milestoneLabel: nil,
                      streakThreshold: nil, streakLabel: nil),
        TrackedMetric(metric: .heartRateVariability, lowerIsBetter: false,
                      milestoneThreshold: 50, milestoneLabel: Copy.Analysis.PersonalRecord.hrvAbove50ms,
                      streakThreshold: nil, streakLabel: nil),
        TrackedMetric(metric: .sleepDuration, lowerIsBetter: false,
                      milestoneThreshold: 8, milestoneLabel: Copy.Analysis.PersonalRecord.eightHourSleepNight,
                      streakThreshold: 7, streakLabel: Copy.Analysis.PersonalRecord.sevenPlusHrSleep),
        TrackedMetric(metric: .restingHeartRate, lowerIsBetter: true,
                      milestoneThreshold: nil, milestoneLabel: nil,
                      streakThreshold: nil, streakLabel: nil),
        TrackedMetric(metric: .standHours, lowerIsBetter: false,
                      milestoneThreshold: nil, milestoneLabel: nil,
                      streakThreshold: 10, streakLabel: Copy.Analysis.PersonalRecord.tenPlusStandHours),
    ]

    /// Analyze personal records, milestones, and streaks
    static func generateInsights(timeSeries: [HealthMetric: MetricTimeSeries]) -> [Insight] {
        var insights: [Insight] = []

        // --- Rolling PRs ---
        insights.append(contentsOf: findRollingPRs(timeSeries: timeSeries))

        // --- Active Streaks ---
        insights.append(contentsOf: findStreaks(timeSeries: timeSeries))

        // --- Milestones ---
        insights.append(contentsOf: findMilestones(timeSeries: timeSeries))

        return insights
    }

    // MARK: - Rolling PRs

    private static func findRollingPRs(timeSeries: [HealthMetric: MetricTimeSeries]) -> [Insight] {
        var insights: [Insight] = []

        for tracked in trackedMetrics {
            guard let series = timeSeries[tracked.metric] else { continue }
            let sorted = series.sortedSamples
            guard sorted.count >= 30 else { continue }

            let values = sorted.map(\.value)

            // 7-day rolling average PR
            if let prInsight = checkRollingPR(
                metric: tracked.metric,
                values: values,
                window: 7,
                windowLabel: "7-day",
                lowerIsBetter: tracked.lowerIsBetter
            ) {
                insights.append(prInsight)
            }

            // 30-day rolling average PR
            if let prInsight = checkRollingPR(
                metric: tracked.metric,
                values: values,
                window: 30,
                windowLabel: "30-day",
                lowerIsBetter: tracked.lowerIsBetter
            ) {
                insights.append(prInsight)
            }
        }

        return insights
    }

    private static func checkRollingPR(
        metric: HealthMetric,
        values: [Double],
        window: Int,
        windowLabel: String,
        lowerIsBetter: Bool
    ) -> Insight? {
        guard values.count >= window else { return nil }

        let movingAverages = values.movingAverage(window: window)
        guard !movingAverages.isEmpty else { return nil }

        guard let currentAvg = movingAverages.last else { return nil }
        let bestPrevious: Double
        if lowerIsBetter {
            bestPrevious = movingAverages.dropLast().min() ?? currentAvg
        } else {
            bestPrevious = movingAverages.dropLast().max() ?? currentAvg
        }

        let isPR: Bool
        if lowerIsBetter {
            isPR = currentAvg <= bestPrevious && movingAverages.count > 1
        } else {
            isPR = currentAvg >= bestPrevious && movingAverages.count > 1
        }

        guard isPR else { return nil }

        let improvement: Double
        if bestPrevious != 0 {
            improvement = abs((currentAvg - bestPrevious) / bestPrevious) * 100
        } else {
            improvement = 0
        }

        // Only report if it's a meaningful PR (within 2% of previous best counts as "tied")
        guard improvement < 2 || movingAverages.count <= window + 1 else {
            // Calculate how long the previous record stood
            let previousBestIndex: Int
            if lowerIsBetter {
                previousBestIndex = movingAverages.dropLast().enumerated().min(by: { $0.element < $1.element })?.offset ?? 0
            } else {
                previousBestIndex = movingAverages.dropLast().enumerated().max(by: { $0.element < $1.element })?.offset ?? 0
            }
            let recordAgeDays = movingAverages.count - 1 - previousBestIndex
            let recordAgeNote = recordAgeDays > 7 ? " You beat a record that stood for \(recordAgeDays) days." : ""

            // It's a new PR with real improvement. use variable templates for copy variety
            let templateVariant = metric.rawValue.count % 3
            let prSummary: String
            let prRecommendation: String
            switch templateVariant {
            case 1:
                prSummary = "New personal best! Your \(windowLabel) \(metric.displayName.lowercased()) just reached \(String(format: "%.1f", currentAvg)) \(metric.unit) \u{2014} that's \(String(format: "%.1f", improvement))% above your previous record.\(recordAgeNote)"
                prRecommendation = "New \(windowLabel) record for \(metric.displayName.lowercased()): \(String(format: "%.1f", currentAvg)) \(metric.unit), surpassing previous best of \(String(format: "%.1f", bestPrevious)) \(metric.unit)."
            case 2:
                prSummary = "Record broken: \(metric.displayName.lowercased()) at \(String(format: "%.1f", currentAvg)) \(metric.unit). Keep this trajectory going.\(recordAgeNote)"
                prRecommendation = "New \(windowLabel) record for \(metric.displayName.lowercased()): \(String(format: "%.1f", currentAvg)) \(metric.unit), surpassing previous best of \(String(format: "%.1f", bestPrevious)) \(metric.unit)."
            default:
                prSummary = "Your \(windowLabel) average \(metric.displayName.lowercased()) hit a personal record: \(String(format: "%.1f", currentAvg)) \(metric.unit). Previous best: \(String(format: "%.1f", bestPrevious)) \(metric.unit) (\(String(format: "%.1f", improvement))% improvement). You're building real momentum.\(recordAgeNote)"
                prRecommendation = "New \(windowLabel) record for \(metric.displayName.lowercased()): \(String(format: "%.1f", currentAvg)) \(metric.unit), surpassing previous best of \(String(format: "%.1f", bestPrevious)) \(metric.unit)."
            }

            return Insight(
                metric: metric,
                title: Copy.Analysis.PersonalRecord.newPR(windowLabel: windowLabel, metricName: metric.displayName),
                summary: prSummary,
                recommendation: prRecommendation,
                severity: .info,
                trend: .improving,
                currentValue: currentAvg,
                baselineValue: bestPrevious,
                deviationPercent: improvement,
                category: .personalRecord,
                relatedMetrics: [metric]
            )
        }

        // First window ever or tied with previous best
        return nil
    }

    // MARK: - Streaks

    private static func findStreaks(timeSeries: [HealthMetric: MetricTimeSeries]) -> [Insight] {
        var insights: [Insight] = []

        for tracked in trackedMetrics {
            guard let threshold = tracked.streakThreshold,
                  let label = tracked.streakLabel,
                  let series = timeSeries[tracked.metric] else { continue }

            let valueMap = TimeSeriesAligner.dailyValueMap(series)
            var currentStreak = 0
            var today = Date().startOfDay

            // Count consecutive days backward from today
            while true {
                if let value = valueMap[today], value >= threshold {
                    currentStreak += 1
                    today = Calendar.current.date(byAdding: .day, value: -1, to: today)?.startOfDay ?? today
                } else {
                    break
                }
            }

            guard currentStreak >= 3 else { continue }

            insights.append(Insight(
                metric: tracked.metric,
                title: Copy.Analysis.PersonalRecord.streakTitle(days: currentStreak, label: label),
                summary: Copy.Analysis.PersonalRecord.streakSummary(label: label, days: currentStreak),
                recommendation: currentStreak >= 7 ?
                    "\(currentStreak)-day streak of \(label.lowercased()). This is your longest active run." :
                    "Current streak: \(currentStreak) consecutive days of \(label.lowercased()).",
                severity: .info,
                trend: .improving,
                currentValue: Double(currentStreak),
                baselineValue: 3,
                deviationPercent: Double(currentStreak - 3) / 3.0 * 100,
                category: .personalRecord,
                relatedMetrics: [tracked.metric]
            ))
        }

        return insights
    }

    // MARK: - Milestones

    private static func findMilestones(timeSeries: [HealthMetric: MetricTimeSeries]) -> [Insight] {
        var insights: [Insight] = []

        for tracked in trackedMetrics {
            guard let threshold = tracked.milestoneThreshold,
                  let label = tracked.milestoneLabel,
                  let series = timeSeries[tracked.metric] else { continue }

            let recent = series.samples(lastDays: 7)
            let achieved = recent.contains { sample in
                tracked.lowerIsBetter ? sample.value <= threshold : sample.value >= threshold
            }

            // Check if this is the first time (not in any older data)
            let older = series.sortedSamples.filter { sample in
                sample.date < Date().daysAgo(7)
            }
            let previouslyAchieved = older.contains { sample in
                tracked.lowerIsBetter ? sample.value <= threshold : sample.value >= threshold
            }

            guard achieved, !previouslyAchieved else { continue }

            insights.append(Insight(
                metric: tracked.metric,
                title: Copy.Analysis.PersonalRecord.milestoneTitle(label),
                summary: Copy.Analysis.PersonalRecord.milestoneSummary(label),
                recommendation: Copy.Analysis.PersonalRecord.milestoneRecommendation(label),
                severity: .info,
                trend: .improving,
                currentValue: threshold,
                baselineValue: 0,
                deviationPercent: 100,
                category: .personalRecord,
                relatedMetrics: [tracked.metric]
            ))
        }

        return insights
    }
}

// MARK: - InsightAnalyzer Conformance

extension PersonalRecordAnalyzer: InsightAnalyzer {
    static var analyzerID: String { "personalRecord" }
    static var insightCategory: InsightCategory { .personalRecord }

    static func generateInsights(context: AnalysisContext) -> [Insight] {
        generateInsights(timeSeries: context.timeSeries)
    }
}

File: android-app/app/src/main/java/com/lasohealth/android/core/analysis/analyzers/PersonalRecordAnalyzer.kt
package com.lasohealth.android.core.analysis.analyzers

import com.lasohealth.android.core.analysis.AnalysisContext
import com.lasohealth.android.core.analysis.AnalysisInsight
import com.lasohealth.android.core.analysis.DailySample
import com.lasohealth.android.core.analysis.InsightAnalyzer
import com.lasohealth.android.core.analysis.InsightDirective
import com.lasohealth.android.core.analysis.InsightPriority
import com.lasohealth.android.core.model.HealthMetric
import java.util.Calendar
import java.util.UUID
import kotlin.math.abs

/**
 * Tracks rolling personal records (7-day / 30-day moving average PRs),
 * activity streaks, and milestone achievements.
 *
 * Ported from iOS PersonalRecordAnalyzer.swift.
 */
class PersonalRecordAnalyzer : InsightAnalyzer {

    override val name: String = "PersonalRecordAnalyzer"

    override suspend fun generateInsights(context: AnalysisContext): List<AnalysisInsight> {
        val insights = mutableListOf<AnalysisInsight>()
        insights += findRollingPRs(context.timeSeries)
        insights += findStreaks(context.timeSeries)
        insights += findMilestones(context.timeSeries)
        return insights
    }

    // ----- Configuration -----

    private data class TrackedMetric(
        val metric: HealthMetric,
        val lowerIsBetter: Boolean,
        val milestoneThreshold: Double?,
        val milestoneLabel: String?,
        val streakThreshold: Double?,
        val streakLabel: String?,
    )

    private val trackedMetrics = listOf(
        TrackedMetric(HealthMetric.STEPS, false, 10_000.0, "10K step day", 8_000.0, "8K+ steps"),
        TrackedMetric(HealthMetric.ACTIVE_CALORIES, false, null, null, null, null),
        TrackedMetric(HealthMetric.EXERCISE_MINUTES, false, null, null, 30.0, "30+ min exercise"),
        TrackedMetric(HealthMetric.VO2_MAX, false, null, null, null, null),
        TrackedMetric(HealthMetric.HEART_RATE_VARIABILITY, false, 50.0, "HRV above 50ms", null, null),
        TrackedMetric(HealthMetric.SLEEP_DURATION, false, 8.0, "8-hour sleep night", 7.0, "7+ hr sleep"),
        TrackedMetric(HealthMetric.RESTING_HEART_RATE, true, null, null, null, null),
        TrackedMetric(HealthMetric.STAND_HOURS, false, null, null, 10.0, "10+ stand hours"),
    )

    // ----- Rolling PRs -----

    private fun findRollingPRs(timeSeries: Map<HealthMetric, List<DailySample>>): List<AnalysisInsight> {
        val insights = mutableListOf<AnalysisInsight>()
        for (tracked in trackedMetrics) {
            val samples = timeSeries[tracked.metric] ?: continue
            val sorted = samples.sortedBy { it.date }
            if (sorted.size < 30) continue
            val values = sorted.map { it.value }

            checkRollingPR(tracked.metric, values, 7, "7-day", tracked.lowerIsBetter)?.let { insights += it }
            checkRollingPR(tracked.metric, values, 30, "30-day", tracked.lowerIsBetter)?.let { insights += it }
        }
        return insights
    }

    private fun checkRollingPR(
        metric: HealthMetric,
        values: List<Double>,
        window: Int,
        windowLabel: String,
        lowerIsBetter: Boolean,
    ): AnalysisInsight? {
        if (values.size < window) return null

        val movingAverages = movingAverage(values, window)
        if (movingAverages.size <= 1) return null

        val currentAvg = movingAverages.last()
        val previous = movingAverages.dropLast(1)
        val bestPrevious = if (lowerIsBetter) previous.min() else previous.max()

        val isPR = if (lowerIsBetter) currentAvg <= bestPrevious else currentAvg >= bestPrevious
        if (!isPR) return null

        val improvement = if (bestPrevious != 0.0) {
            abs((currentAvg - bestPrevious) / bestPrevious) * 100.0
        } else 0.0

        // Only report meaningful PRs (>2% improvement) unless it's the first window
        if (improvement < 2.0 && movingAverages.size > window + 1) return null

        val previousBestIndex = if (lowerIsBetter) {
            previous.withIndex().minByOrNull { it.value }?.index ?: 0
        } else {
            previous.withIndex().maxByOrNull { it.value }?.index ?: 0
        }
        val recordAgeDays = previous.size - previousBestIndex
        val recordAgeNote = if (recordAgeDays > 7) " You beat a record that stood for $recordAgeDays days." else ""

        return AnalysisInsight(
            id = UUID.randomUUID().toString(),
            title = "New $windowLabel PR: ${metric.displayName}",
            detail = "Your $windowLabel average ${metric.displayName.lowercase()} hit a personal record: " +
                "${fmt1(currentAvg)} ${metric.unit}. Previous best: ${fmt1(bestPrevious)} ${metric.unit} " +
                "(${fmt1(improvement)}% improvement).$recordAgeNote",
            metric = metric,
            priority = InsightPriority.LOW,
            directive = InsightDirective.MAINTAIN,
        )
    }

    // ----- Streaks -----

    private fun findStreaks(timeSeries: Map<HealthMetric, List<DailySample>>): List<AnalysisInsight> {
        val insights = mutableListOf<AnalysisInsight>()
        for (tracked in trackedMetrics) {
            val threshold = tracked.streakThreshold ?: continue
            val label = tracked.streakLabel ?: continue
            val samples = timeSeries[tracked.metric] ?: continue

            val valueMap = buildDayValueMap(samples)
            var currentStreak = 0
            var day = startOfToday()

            while (true) {
                val value = valueMap[day]
                if (value != null && value >= threshold) {
                    currentStreak++
                    day -= DAY_MS
                } else {
                    break
                }
            }

            if (currentStreak < 3) continue

            insights += AnalysisInsight(
                id = UUID.randomUUID().toString(),
                title = "$currentStreak-Day Streak: $label",
                detail = "You've hit $label for $currentStreak consecutive days. Keep the momentum going!",
                metric = tracked.metric,
                priority = InsightPriority.INFORMATIONAL,
                directive = InsightDirective.MAINTAIN,
            )
        }
        return insights
    }

    // ----- Milestones -----

    private fun findMilestones(timeSeries: Map<HealthMetric, List<DailySample>>): List<AnalysisInsight> {
        val insights = mutableListOf<AnalysisInsight>()
        val now = System.currentTimeMillis()

        for (tracked in trackedMetrics) {
            val threshold = tracked.milestoneThreshold ?: continue
            val label = tracked.milestoneLabel ?: continue
            val samples = timeSeries[tracked.metric] ?: continue

            val recent = samples.filter { it.date > now - 7 * DAY_MS }
            val achieved = recent.any { sample ->
                if (tracked.lowerIsBetter) sample.value <= threshold else sample.value >= threshold
            }

            val older = samples.filter { it.date > now - 90 * DAY_MS && it.date <= now - 7 * DAY_MS }
            val previouslyAchieved = older.any { sample ->
                if (tracked.lowerIsBetter) sample.value <= threshold else sample.value >= threshold
            }

            if (!achieved || previouslyAchieved) continue

            insights += AnalysisInsight(
                id = UUID.randomUUID().toString(),
                title = "Milestone: $label",
                detail = "You reached a new milestone: $label for the first time in 90 days. " +
                    "This shows real progress.",
                metric = tracked.metric,
                priority = InsightPriority.LOW,
                directive = InsightDirective.MAINTAIN,
            )
        }
        return insights
    }

    // ----- Helpers -----

    private fun movingAverage(values: List<Double>, window: Int): List<Double> {
        if (values.size < window) return emptyList()
        return (window..values.size).map { end ->
            values.subList(end - window, end).average()
        }
    }

    private fun buildDayValueMap(samples: List<DailySample>): Map<Long, Double> {
        val map = mutableMapOf<Long, Double>()
        for (sample in samples) {
            val day = startOfDay(sample.date)
            map[day] = sample.value // last write wins for same day
        }
        return map
    }

    private fun startOfDay(epochMs: Long): Long {
        val cal = Calendar.getInstance().apply { timeInMillis = epochMs }
        cal.set(Calendar.HOUR_OF_DAY, 0)
        cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0)
        cal.set(Calendar.MILLISECOND, 0)
        return cal.timeInMillis
    }

    private fun startOfToday(): Long = startOfDay(System.currentTimeMillis())

    private fun fmt1(v: Double): String = String.format("%.1f", v)

    companion object {
        private const val DAY_MS = 86_400_000L
    }
}
